### ISSUE TITLE
Cannot handle types defined in Flutter dart:ui

### ISSUE THREAD

**Issue by diablodale** [dart-lang/json_serializable#236]
**UPDATE**: Likely due to and blocked by https://github.com/dart-lang/build/issues/733

`Color` fields with `@JsonKey(fromJson: _colorFromJson, toJson: _colorToJson)` annotations fail.

## Setup

* json_serializable 0.5.7
* json_annotation 0.2.8
* flutter 0.4.4

## Repo

1. Use the following code

```dart
@JsonSerializable(nullable: true, includeIfNull: false)
class TestClass extends Object with _$TestClassSerializerMixin {
  @JsonKey(fromJson: _colorFromJson, toJson: _colorToJson)
  Color color;

  TestClass();
  factory TestClass.fromJson(Map<String, dynamic> jsonMap) => _$TestClassFromJson(jsonMap);
}

Color _colorFromJson(String colorString) {
  int intColor = int.tryParse(colorString, radix: 16);
  if (intColor == null)
    return null;
  else
    return new Color(intColor);
}

String _colorToJson(Color color) => color.value.toRadixString(16);
```

2. Run `flutter packages pub run build_runner build`

## Result

Build fails. No usable xxx.g.dart file created. Errors on screen.

```
[SEVERE] json_serializable on lib/data/xxx.dart:
Error running JsonSerializableGenerator
At least one field has an invalid type: `color`.
package:xxxxxxxx/data/xxxx.dart:63:9
  Color color;
        ^^^^^
```

## Comments

Seems related to functionality in https://github.com/dart-lang/json_serializable/issues/133#issuecomment-387906489 by @kevmoo .

**Comment by kevmoo** [comment]
That is super weird.

1) Have you run `pub upgrade` lately?
2) What is the Dart version when your run `flutter --version` (after `Tools).

**Comment by diablodale** [comment]
1. Yes. Ran it again now for goodness. All was already up to date.

2.  
```
Flutter 0.4.4 • channel beta • https://dalep-cli@github.com/flutter/flutter.git
Framework • revision f9bb4289e9 (5 weeks ago) • 2018-05-11 21:44:54 -0700
Engine • revision 06afdfe54e
Tools • Dart 2.0.0-dev.54.0.flutter-46ab040e58
```

**Comment by kevmoo** [comment]
Try this

```console
> flutter packages pub run build_runner clean
> flutter packages pub run build_runner build --verbose
```

**Comment by kevmoo** [comment]
...and paste the output

**Comment by diablodale** [comment]
Sure.
```
flutter packages pub run build_runner clean

[INFO] Generating build script...
[INFO] Generating build script completed, took 4.0s

[WARNING] Deleting cache and generated source files.
This shouldn't be necessary for most applications, unless you have made intentional edits to generated files (i.e. for testing). Consider filing a bug at https://github.com/dart-lang/build/issues/new if you are using this to work around an apparent (and reproducible) bug.
[INFO] Cleaning up source outputs...
[INFO] Cleaning up source outputs completed, took 1.3s

[INFO] Cleaning up cache directory...
[INFO] Cleaning up cache directory completed, took 13ms
```

```
flutter packages pub run build_runner build --verbose

[INFO] Generating build script...
[INFO] Generating build script completed, took 4.0s

[INFO] BuildDefinition:Initializing inputs
[INFO] BuildDefinition:Building new asset graph...
[INFO] BuildDefinition:Building new asset graph completed, took 2.1s

[INFO] BuildDefinition:Checking for unexpected pre-existing outputs....
[INFO] BuildDefinition:Checking for unexpected pre-existing outputs. completed, took 24ms

[INFO] Build:Running build...
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[INFO] Heartbeat:12.2s elapsed, 3/20 actions completed.
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[INFO] Heartbeat:15.0s elapsed, 6/20 actions completed.
[INFO] Heartbeat:16.2s elapsed, 6/20 actions completed.
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[INFO] Heartbeat:20.0s elapsed, 6/20 actions completed.
[INFO] Heartbeat:21.1s elapsed, 10/20 actions completed.
[SEVERE] json_serializable on lib/notaproblemredacted.dart:
Unable to resolve asset ID for "dart:ui"
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on test/widget_test.dart:Running JsonSerializableGenerator - 1 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on test/widget_test.dart:Running JsonLiteralGenerator - 2 of 2
[SEVERE] json_serializable on lib/redacted.dart:
Error running JsonSerializableGenerator
At least one field has an invalid type: `color`.
package:redactedproblemfile.dart:48:9
  Color color;
        ^^^^^
package:json_serializable/src/utils.dart 148:5                          createSortedFieldSet
package:json_serializable/src/generator_helper.dart 94:24               _GeneratorHelper._writeCtor
package:json_serializable/src/generator_helper.dart 67:28               _GeneratorHelper._generate
package:json_serializable/src/generator_helper.dart 32:17               generate
package:json_serializable/src/json_serializable_generator.dart 159:7    JsonSerializableGenerator.generateForAnnotatedElement
package:source_gen/src/generator_for_annotation.dart 40:15              GeneratorForAnnotation.generate.<fn>
dart:async                                                              Future.wait
package:source_gen/src/generator_for_annotation.dart 39:34              GeneratorForAnnotation.generate
package:source_gen/src/builder.dart 214:35                              _generate
package:source_gen/src/builder.dart 76:15                               _Builder._generateForLibrary
package:source_gen/src/builder.dart 67:11                               _Builder.build
package:build                                                           runBuilder
package:build_runner_core/src/generate/build_impl.dart 424:15           _SingleBuild._runForInput.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 246:73  _NoOpBuilderActionTracker.track
package:build_runner_core/src/generate/build_impl.dart 423:19           _SingleBuild._runForInput
package:build_runner_core/src/generate/build_impl.dart 349:38           _SingleBuild._runBuilder.<fn>
dart:async                                                              Future.wait
package:build_runner_core/src/generate/build_impl.dart 348:36           _SingleBuild._runBuilder
package:build_runner_core/src/generate/build_impl.dart 296:32           _SingleBuild._runPhases.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 140:15  _NoOpBuildPerformanceTracker.trackBuildPhase
package:build_runner_core/src/generate/build_impl.dart 292:33           _SingleBuild._runPhases
package:build_runner_core/src/logging/logging.dart 25:30                logTimedAsync
package:build_runner_core/src/generate/build_impl.dart 260:26           _SingleBuild._safeBuild.<fn>
dart:async                                                              runZoned
package:build_runner_core/src/generate/build_impl.dart 255:5            _SingleBuild._safeBuild
package:build_runner_core/src/generate/build_impl.dart 185:24           _SingleBuild.run
package:build_runner_core/src/generate/build_impl.dart 89:30            BuildImpl.run
package:build_runner_core/src/generate/build_runner.dart 23:14          BuildRunner.run
package:build_runner                                                    BuildCommand.run
package:args/command_runner.dart 194:27                                 CommandRunner.runCommand
package:args/command_runner.dart 109:29                                 CommandRunner.run.<fn>
dart:async                                                              new Future.sync
package:args/command_runner.dart 109:11                                 CommandRunner.run
package:build_runner                                                    run
.dart_tool\build\entrypoint\build.dart 11:22                            main

[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[FINE] json_serializable on lib/redacted.dart:Running JsonLiteralGenerator - 2 of 2
[INFO] Build:Running build completed, took 34.5s

[INFO] Build:Caching finalized dependency graph...
[INFO] Build:Caching finalized dependency graph completed, took 305ms

[INFO] Build:Succeeded after 34.9s with 6 outputs (20 actions)
```

**Comment by kevmoo** [comment]
Where is the `Color` type defined? Is it in the current package or somewhere else?

**Comment by diablodale** [comment]
`Color` is a core Flutter library type.

At the top of the file having this error, I have the line `import 'package:flutter/material.dart';`

Somewhere within that core Flutter package it leads to the Color definition in `<fluttersdkdir>\bin\cache\pkg\sky_engine\lib\ui\painting.dart`

**Comment by kevmoo** [comment]
I bet you anything that's the issue. Will look into it. Thanks!

**Comment by diablodale** [comment]
Thanks for considering. I poked around myself, but haven't gotten my head around everything needed to address it.

I did also look at type helpers and thought of writing a custom one for `Color` for my app to perhaps get around this issue and enable it globally. Though, I again haven't yet enough combined knowledge of build, dart, and this package to write it and enable it. Though, the helper code itself, like in the recent helper for Uri, doesn't look that difficult.

**Comment by kevmoo** [comment]
@diablodale look up the log you sent

```
[SEVERE] json_serializable on lib/notaproblemredacted.dart:
Unable to resolve asset ID for "dart:ui"
```

Looking into this now...

**Comment by kevmoo** [comment]
I hacked into `build_resolver` to add a stack there.

@jakemac53 Could this be a bug in build_resolvers? Or pkg:analyzer?

```
[SEVERE] json_serializable on lib/example.dart:
Unable to resolve asset ID for "dart:ui"
oops
package:build_resolvers/src/resolver.dart 292:24                        _AssetUriResolver.resolveAbsolute
package:analyzer/src/context/source.dart 259:34                         SourceFactoryImpl._internalResolveUri.<fn>
dart:collection                                                         _HashMap.putIfAbsent
package:analyzer/src/context/source.dart 257:38                         SourceFactoryImpl._internalResolveUri
package:analyzer/src/context/source.dart 158:14                         SourceFactoryImpl.resolveUri
package:analyzer/src/task/dart.dart 3944:36                             ParseDartTask._resolveUri
package:analyzer/src/task/dart.dart 3897:23                             ParseDartTask._resolveDirective
package:analyzer/src/task/dart.dart 3797:37                             ParseDartTask.internalPerform
package:analyzer/src/task/api/model.dart 323:9                          AnalysisTask._safelyPerform
package:analyzer/src/task/api/model.dart 222:7                          AnalysisTask.perform
package:analyzer/src/task/driver.dart 285:10                            AnalysisDriver.performWorkItem
package:analyzer/src/task/driver.dart 110:22                            AnalysisDriver.computeResult
package:analyzer/src/context/context.dart 722:14                        AnalysisContextImpl.computeResult
package:analyzer/src/context/context.dart 680:14                        AnalysisContextImpl.computeKindOf
package:build_resolvers/src/resolver.dart 99:16                         AnalyzerResolver._isLibrary
package:build_resolvers/src/resolver.dart 95:30                         AnalyzerResolver.isLibrary
package:build_resolvers/src/resolver.dart 61:56                         PerActionResolver.isLibrary
package:build                                                           _DelayedResolver.isLibrary
package:source_gen/src/builder.dart 65:25                               _Builder.build
package:build                                                           runBuilder
package:build_runner_core/src/generate/build_impl.dart 424:15           _SingleBuild._runForInput.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 246:73  _NoOpBuilderActionTracker.track
package:build_runner_core/src/generate/build_impl.dart 423:19           _SingleBuild._runForInput
package:build_runner_core/src/generate/build_impl.dart 349:38           _SingleBuild._runBuilder.<fn>
dart:async                                                              Future.wait
package:build_runner_core/src/generate/build_impl.dart 348:36           _SingleBuild._runBuilder
package:build_runner_core/src/generate/build_impl.dart 296:32           _SingleBuild._runPhases.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 140:15  _NoOpBuildPerformanceTracker.trackBuildPhase
package:build_runner_core/src/generate/build_impl.dart 292:33           _SingleBuild._runPhases
package:build_runner_core/src/logging/logging.dart 25:30                logTimedAsync
package:build_runner_core/src/generate/build_impl.dart 260:26           _SingleBuild._safeBuild.<fn>
dart:async                                                              runZoned
package:build_runner_core/src/generate/build_impl.dart 255:5            _SingleBuild._safeBuild
package:build_runner_core/src/generate/build_impl.dart 185:24           _SingleBuild.run
package:build_runner_core/src/generate/build_impl.dart 89:30            BuildImpl.run
package:build_runner                                                    WatchCommand.run
package:args/command_runner.dart 194:27                                 CommandRunner.runCommand
package:args/command_runner.dart 109:29                                 CommandRunner.run.<fn>
dart:async                                                              new Future.sync
package:args/command_runner.dart 109:11                                 CommandRunner.run
package:build_runner                                                    run
.dart_tool/build/entrypoint/build.dart 11:22                            main
```

**Comment by diablodale** [comment]
That is a known issue and in a separate file. It causes no harm of which I'm aware.
For example here https://github.com/google/built_value.dart/issues/282

**Comment by kevmoo** [comment]
Here's the crux, it seems - https://github.com/dart-lang/build/issues/733

**Comment by diablodale** [comment]
That is a known issue. As in https://github.com/google/built_value.dart/issues/282

That specific error has always been there for me. And it caused no harm to any generated files. I can fully use json_serializable and ignore that error. FYI, also that error comes from a separate dart file than the Color example I gave.

Ahh, I see you typing above. This *is* the first time that I have defining a custom to/fromJson handler for a Flutter type. Other times I have used types outside Flutter.

**Comment by diablodale** [comment]
I might get something similar to work by defining a private field and then custom get/setters. I did that with DateTime because I wanted my DateTime to appear differently than the current DateTime typehandler.

**Comment by kevmoo** [comment]
That'd likely be a good way to handle this...

**Comment by diablodale** [comment]
Workaround fail. :-/ I've found that `@JsonKey(ignore: true)` isn't fully ignoring fields and methods. 
```dart
@JsonKey(ignore: true)
Color color;

@JsonKey(name: "color")
String get colorString => _colorToJson(this.color);
@JsonKey(name: "color")
set colorString(String value) => _colorFromJson(value);
```
It continues to result in the same error `At least one field has an invalid type: color` error.
Perhaps this is a separate issue with the `ignore` annotation exposed through my workaround attempt?

I even tried a workaround^2. It failed with similar `At least one field has an invalid type: _color, color.`
```dart
@JsonKey(ignore: true)
Color _color;

@JsonKey(ignore: true)
Color get color => this._color;
@JsonKey(ignore: true)
set color(Color value) => this._color = value;

@JsonKey(name: "color")
String get colorString => _colorToJson(this._color);
@JsonKey(name: "color")
set colorString(String value) => this._color = _colorFromJson(value);
```

**Comment by kevmoo** [comment]
I have a plan to fix this here. Thanks for your patience...

**Comment by kevmoo** [comment]
PR out w/ a fix (work-around) https://github.com/dart-lang/json_serializable/pull/243

**Comment by diablodale** [comment]
Tested https://github.com/dart-lang/json_serializable/pull/243 which was included in v0.5.8 with my scenario of Color...

I get generated code that works :-)  Thanks!

Two oddities:

1. I continue to get warnings during build runner on that `Color` field.
2. In the generated mixin code, that field has a getter with `dynamic` rather than `Color`

```dart
abstract class _$MyObjectSerializerMixin {
  dynamic get color;
```

**Comment by kevmoo** [comment]
Great to hear it works. Both oddities are by design. #1 is just being super
clear that the type is weird. #2 is because – well – there's no way to do
better. ;-)

**Comment by kevmoo** [comment]
Would you run w/ `--verbose` so I can see the stack trace?

**Comment by diablodale** [comment]
In current working state of older versions, won't be helpful. I would need to reinstall the failing package versions.
Though what you see in the two blocks above had been flutter and buildrunner cleaned. I did that as part of my troubleshooting before coming back to this bug.

To be clear are you asking the following?

1. build_runner clean (I've discovered this needs to be done before switching versions, otherwise, the clean of the new version faults one its own)
2. flutter clean
3. go back to the failing versions
4. build_runner clean
5. build_runner build --verbose

**Comment by kevmoo** [comment]
If possible, please.

**Comment by kevmoo** [comment]
Figured it out

1) unfound assets cause severe warnings

https://github.com/dart-lang/build/blob/81082de9aed7e4af4945de6d0381d09198593465/build_resolvers/lib/src/resolver.dart#L299-L302

2) build_runner now fails builds with severe log entries - as of ^0.9.0

https://github.com/dart-lang/build/commit/c5513e2

So we need to fix one of those. I'd probably vote for fixing the first.

Thoughts, @natebosch @jakemac53 ?

**Comment by jakemac53** [comment]
See the assert at the top of that method.... https://github.com/dart-lang/build/blob/81082de9aed7e4af4945de6d0381d09198593465/build_resolvers/lib/src/resolver.dart#L292

We aren't expecting to get called for any `dart:` uris here. My best guess is that this is actually an analyzer bug, where it doesn't invoke that method for any other `dart:` uris but does for this one somehow?

**Comment by kevmoo** [comment]
So https://github.com/dart-lang/build/pull/1776 fixes things – need to discuss w/ build team if this is a reasonable fix...

**Comment by kevmoo** [comment]
@diablodale – do a `flutter packages upgrade` and try again - you should be good!

**Comment by diablodale** [comment]
Yes, it is again generating the needed xxx.g.dart code.

FYI, some of the package versions that changed and seem relevant in this new working scenario are:
build_resolvers 0.2.2+3
build_runner 0.10.1.+1
build_runner_core 0.3.1+4
json_annotation 1.1.0
json_serializable 1.1.0
source_gen 0.9.0+1

I consider the workaround again in place and working.
I have not tested the original scenario that does not have this `@JsonKey from/toJson` workaround.

**Comment by pavanpodila** [comment]
This is again failing when using with `mobx_codegen`. Doesn't throw any errors but I get **`dynamic`** as the discovered type for `dart:ui.Color`

Issue: https://github.com/mobxjs/mobx.dart/issues/165

```
$>  flutter packages pub run build_runner build --verbose



[INFO] Generating build script...lutter packages pub run build_runner build --verbose
[INFO] Generating build script completed, took 350ms

[INFO] Creating build script snapshot......
[INFO] Creating build script snapshot... completed, took 10.3s

[INFO] BuildDefinition:Initializing inputs
[INFO] BuildDefinition:Building new asset graph...
[INFO] BuildDefinition:Building new asset graph completed, took 682ms

[INFO] BuildDefinition:Checking for unexpected pre-existing outputs....
[INFO] BuildDefinition:Checking for unexpected pre-existing outputs. completed, took 2ms

[INFO] Build:Running build...
[FINE] mobx_codegen:mobx_generator on lib/multi_counter/multi_counter_store.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/counter/counter.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/todos/todo.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/todos/todo_list.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/form/form_store.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/clock/clock.dart:Running StoreGenerator
[INFO] Heartbeat:6.1s elapsed, 6/18 actions completed.
[FINE] mobx_codegen:mobx_generator on lib/form/form_widgets.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/multi_counter/multi_counter_widgets.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/counter/counter_widgets.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/todos/todo_widgets.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/clock/clock_widgets.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/github/github_store.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/github/github_widgets.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/examples.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/main.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/hackernews/news_store.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on lib/hackernews/news_widgets.dart:Running StoreGenerator
[FINE] mobx_codegen:mobx_generator on test/widget_test.dart:Running StoreGenerator
[INFO] Build:Running build completed, took 8.3s

[INFO] Build:Caching finalized dependency graph...
[INFO] Build:Caching finalized dependency graph completed, took 56ms

[INFO] Build:Succeeded after 8.4s with 14 outputs (43 actions)

```

**Comment by kevmoo** [comment]
Sadly, changes in `pkg:analyzer` lead to this. Specifically, with the latest release `DartType.isUndefined` now always returns `false`. This is the case being hit w/ Flutter.

To quote @stereotype441 

> Long term our plan is to add an InvalidType similar to what CFE does (see https://github.com/dart-lang/sdk/issues/36697).  It doesn't make sense to have both InvalidType and UndefinedType, and we decided that to ease the analyzer maintenance burden it would be better to remove UndefinedType now.  So yes, it means that clients using isUndefined are currently out of luck.

**Comment by mrruby** [comment]
After upgrading to 2.3.0 its fix for me :)

**Comment by BharathMG** [comment]
This issue is still present in following version,

```
Flutter 1.7.8+hotfix.4 • channel beta • https://github.com/flutter/flutter.git
Framework • revision 20e59316b8 (5 days ago) • 2019-07-18 20:04:33 -0700
Engine • revision fee001c93f
Tools • Dart 2.4.0
```

If I use any `enum` from `dart:ui` package, `build_runner` **does not resolve them to Enums**. They are instead just getting converted to `json[field]` as `String`. Is it a known issue?

**Comment by kevmoo** [comment]
@jakemac53 @natebosch – could either of you folks take a look?

**Comment by natebosch** [comment]
This should work if you have `build_resolvers` version `1.2.1` or greater.

**Comment by kika** [comment]
@natebosch My project uses `build_resolvers` `1.2.1` according to `pubspec.lock` and still `StrokeCap` from `dart:ui` is not recognized as an enum.

**Comment by jakemac53** [comment]
Are the existing issues specific to `Enum`s?

**Comment by kevmoo** [comment]
`dart:ui` types are visible to the builder. In general we just "let things
through" but I'm guessing this is failing because we cannot "see" that it's
an enum

**Comment by ksilz** [comment]
Issue is still here when trying to serialized a `Color`. Using `json_serializable 3.4.1` through `freezed 0.12.1` with Flutter 1.20.4:

```
Could not generate `fromJson` code for `iosPrimaryColor`.
None of the provided `TypeHelper` instances support the defined type.
package:y_mobile/redux/state/display_colors.freezed.dart:160:15
    ╷
160 │   final Color iosPrimaryColor;
    │               ^^^^^^^^^^^^^^^
```

**Comment by kevmoo** [comment]
@ksilz – did you do any work to create a converter for this field? `Color` doesn't have serialization support baked in!

**Comment by ksilz** [comment]
@kevmoo I changed the fields to `int`. Then I added getters that return `Color`. May make it more efficient later.

**Comment by SalahAdDin** [comment]
> @ksilz – did you do any work to create a converter for this field? `Color` doesn't have serialization support baked in!

Is it required to create a custom converter? Cause it was working fine before i update the package.

**Comment by huang12zheng** [comment]
// source from: https://developer.school/tutorials/flutter-using-json_serializable-to-serialise-dart-classes
/// ```dart
/// @ColorSerialiser()
/// final Color favouriteColour;
/// ```
searching for solve it. @SalahAdDin
